### ISSUE TITLE
About the result

### ISSUE THREAD

**Issue by bowen-xiao96** [grey-eye/talking-heads#25]
Hello! I am also trying to reproduce this paper. However, when using very similar network architecture, I can only achieve relatively low-resolution faces placed on fuzzy background. Your github code seems to be not working either. I would like to ask whether you have achieved better results, and what the core trick is. Thank you!

**Comment by MrCaracara** [comment]
Hi @bowen-xiao96, 

I haven't been able to achieve real results either yet, so I don't know what is missing. I also checked out a version of my code someone else [posted](https://github.com/vincent-thevenin/Realistic-Neural-Talking-Head-Models/issues), but it doesn't seem to work either. 
I get some face-like creations with my network after 2 epochs on a 120000 subset of the entire dataset, but it's not acceptable yet.

It could be that the network simply needs more training time.

**Comment by bowen-xiao96** [comment]
Hi,
Here are the results that I have achieved.
![1](https://user-images.githubusercontent.com/39450130/62916248-30b93b80-bdca-11e9-8d75-403c6edbe24a.png)
![2](https://user-images.githubusercontent.com/39450130/62916249-30b93b80-bdca-11e9-9f35-35b78827b53d.png)
![3](https://user-images.githubusercontent.com/39450130/62916250-30b93b80-bdca-11e9-96ed-6aae6a4d11fe.png)

I followed all the details in the original paper. I used all the loss functions, and adopted the same residual block structure as BigGAN. Notice that I did not put any activation after the output of the discriminator.

I used VoxCeleb1 dataset, which includes 2w videos. The network was trained for 25 epoches with a batch size of 3. Although the network can capture some outlines of the person, the generated images are fuzzy and has strong checkerboard artifacts. The loss of the discriminator can converge to 2, which is the optimal value. However the content loss is still high. I believe the problem lies in the embedder. I will try to run on VoxCeleb2 with more epoches, and will give a result of finetuning.

**Comment by bowen-xiao96** [comment]
I found two tricks in network training. The first is that the content loss should be taken from the output of the convolution (rather than after relu activation). This can enhance the details of the generated images. The other one is that the image should be properly normalized before entering the two VGGs. I directly used caffe vgg model so the input range should be [0, 255] plus imagenet mean subtraction.

**Comment by castelo-software** [comment]
> I found two tricks in network training. The first is that the content loss should be taken from the output of the convolution (rather than after relu activation). This can enhance the details of the generated images. The other one is that the image should be properly normalized before entering the two VGGs. I directly used caffe vgg model so the input range should be [0, 255] plus imagenet mean subtraction.

Hi @bowen-xiao96,

These results are very similar to the ones that we're getting now, so it seems that the differences between your implementation and the one in the master branch may not have a big effect. Unless yours converges faster. In which epoch for how many training examples are you with these results?
I believe that this checkerboard pattern around the face is just a matter of not having trained long enough.

As for the normalization, what do you mean that "the image should be properly normalized"? In my code I'm already doing ImgNet normalization.
https://github.com/grey-eye/talking-heads/blob/master/network/loss.py#L27

**Comment by bowen-xiao96** [comment]
Here are the results after finetuning. Finetuning seems to be the final crack.
![0](https://user-images.githubusercontent.com/39450130/62945759-f7a4b980-be11-11e9-916c-fe7ddbe91967.jpg)
![1](https://user-images.githubusercontent.com/39450130/62945763-f83d5000-be11-11e9-907d-cf77b88d84bb.jpg)
![2](https://user-images.githubusercontent.com/39450130/62945766-f8d5e680-be11-11e9-8c15-2452453c9a93.jpg)
![3](https://user-images.githubusercontent.com/39450130/62945773-fa071380-be11-11e9-9239-ae70164006aa.jpg)

I trained the network on the VoxCeleb1 dataset (2.2w videos with 50 frames sampled from each video). Since the batch size is 3, in each epoch there are about 7k steps. The results are obtained after 25 epoches of training. However the content loss got saturated in about 15 epoches, so I believe it will not become further better. I will now train the model on the full VoxCeleb2 and see.

**Comment by castelo-software** [comment]
Hey @bowen-xiao96 !
I still haven't gotten around to implementing fine tuning. I was going to do that now, but I found out there's an error that occurs sporadically in my model:
When putting the data through the AdaIN layers, sometimes the gradient of the standard deviation goes to NaN, possibly because I get a square root of a number equal to or smaller than 0.
This would probably be solved using ReLU's in the Generator, like I had at some point.

So I'm curious about how you have implemented the Generator. Specifically: where and how many ReLU's have you placed? Did you use IN in the Generator like the code is right now?

**Comment by ghost** [comment]
@bowen-xiao96 great results are you planning to release your pretrained models?

**Comment by mrgloom** [comment]
@bowen-xiao96 I wonder what is the source of the problem of "wavy checkerboard" artefacts on hair and shoulders?

**Comment by bowen-xiao96** [comment]
> Hey @bowen-xiao96 !
> I still haven't gotten around to implementing fine tuning. I was going to do that now, but I found out there's an error that occurs sporadically in my model:
> When putting the data through the AdaIN layers, sometimes the gradient of the standard deviation goes to NaN, possibly because I get a square root of a number equal to or smaller than 0.
> This would probably be solved using ReLU's in the Generator, like I had at some point.
> 
> So I'm curious about how you have implemented the Generator. Specifically: where and how many ReLU's have you placed? Did you use IN in the Generator like the code is right now?

Hi @MrCaracara!
My network is not as deep as yours. It only downsamples the image to 16x16, and it has no additional nn.InstanceNorm2d in the decoding part. I also did not put any relu at the end of any residual block. I believe other parts are basically the same, so I have no idea what accounts for the numerical instability.

**Comment by bowen-xiao96** [comment]
> @bowen-xiao96 great results are you planning to release your pretrained models?

I plan to release the code after some minor changes. Since the model on voxceleb1 is not so satisfactory, I am now training it on voxceleb2. I will release the model when all is complete.

**Comment by bowen-xiao96** [comment]
> @bowen-xiao96 I wonder what is the source of the problem of "wavy checkerboard" artefacts on hair and shoulders?

A major source of the checkerboard artifact comes from deconvolution layers. Since there is no deconvolution layer in the network, I doubt whether nearest neighbor upsampling is the problem. I am now testing the more accurate bilinear upsampling.

**Comment by mrgloom** [comment]
@bowen-xiao96 

But I wonder why artifacts are appear in "uncertainty" areas(hair and shoulder), i.e. face looks blurry, but ok.

If it's just artifact introduced by nearest neighbor upsampling
https://github.com/grey-eye/talking-heads/blob/35a1e1c3c4ab79fe5d17a14a744b4fd2642f4fcc/network/components.py#L116
Maybe it can be fixed just adding 1-2 extra conv layers just after upsampling layer?

**Comment by bowen-xiao96** [comment]
> @bowen-xiao96
> 
> But I wonder why artifacts are appear in "uncertainty" areas(hair and shoulder), i.e. face looks blurry, but ok.
> 
> If it's just artifact introduced by nearest neighbor upsampling
> https://github.com/grey-eye/talking-heads/blob/35a1e1c3c4ab79fe5d17a14a744b4fd2642f4fcc/network/components.py#L116
> 
> 
> Maybe it can be fixed just adding 1-2 extra conv layers just after upsampling layer?

These hair and shoulder areas are of high frequency, compared with face which has a large area of constant color. I also doubt my network has not converged. Below is the graph of the content loss with a sigma=10 Gaussian filter applied. However, doing more training does not seem to make it go down more.
![Figure_1](https://user-images.githubusercontent.com/39450130/63315117-9d898400-c33c-11e9-84a1-8ddc004fda04.png)

**Comment by mrgloom** [comment]
As a side not in FSGAN paper https://arxiv.org/pdf/1908.05932.pdf they also have similar artifacts on hair and shoulder.

**Comment by mikirui** [comment]
@bowen-xiao96 It would be great help for us if you can release your codes! Many thanks~

**Comment by visonpon** [comment]
@bowen-xiao96 maybe the landmarks' outline of the face help to generate the better face than shoulder and hair. you can use the canny algorithm to detect the edge of hair and shoulder ,  and see if this help~

**Comment by mrgloom** [comment]
>canny algorithm to detect the edge of hair and shoulder

Maybe it can help, but It should be unstable between frames and produce jitter artifacts at generation time.

Here they seems use some additional losses to force consistency between frames.
https://github.com/NVIDIA/vid2vid
https://raw.githubusercontent.com/NVIDIA/vid2vid/master/imgs/face.gif
https://raw.githubusercontent.com/NVIDIA/vid2vid/master/imgs/face_multiple.gif

**Comment by zhuhaozh** [comment]
@bowen-xiao96 looking forward to your codes! Is there any updates?
